File: web/src/features/system-settings/general/__tests__/contact-qr-code.test.ts
/*
Copyright (C) 2023-2026 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/
import { expect, test } from 'vitest'

import {
  getContactQRCodeUploadError,
  isContactQRCodeValue,
  maxContactQRCodeSize,
} from '../contact-qr-code'

test('accepts supported contact QR code images within the size limit', () => {
  expect(
    getContactQRCodeUploadError({
      type: 'image/png',
      size: maxContactQRCodeSize,
    })
  ).toBeUndefined()
})

test('rejects unsupported or oversized contact QR code images', () => {
  expect(
    getContactQRCodeUploadError({ type: 'image/svg+xml', size: 128 })
  ).toBe('invalid_type')
  expect(
    getContactQRCodeUploadError({
      type: 'image/webp',
      size: maxContactQRCodeSize + 1,
    })
  ).toBe('too_large')
})

test('accepts uploaded QR code paths and HTTPS URLs only', () => {
  expect(
    isContactQRCodeValue(
      '/api/contact/qr-code/15b2e26fd23642f08725765dfe16f042'
    )
  ).toBe(true)
  expect(isContactQRCodeValue('https://example.com/contact-qr.png')).toBe(true)
  expect(isContactQRCodeValue('data:image/png;base64,AAAA')).toBe(false)
})

File: controller/contact_qr_code_test.go
/*
Copyright (C) 2023-2026 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/
package controller

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestContactQRCodeContentTypeRecognizesSupportedImageHeaders(t *testing.T) {
	tests := []struct {
		name        string
		data        []byte
		contentType string
		valid       bool
	}{
		{
			name:        "png",
			data:        []byte{0x89, 'P', 'N', 'G', '\r', '\n', 0x1a, '\n'},
			contentType: "image/png",
			valid:       true,
		},
		{
			name:        "jpeg",
			data:        []byte{0xff, 0xd8, 0xff},
			contentType: "image/jpeg",
			valid:       true,
		},
		{
			name:        "webp",
			data:        []byte("RIFF\x00\x00\x00\x00WEBP"),
			contentType: "image/webp",
			valid:       true,
		},
		{
			name:  "unsupported",
			data:  []byte("not an image"),
			valid: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			contentType, valid := contactQRCodeContentType(tt.data)
			assert.Equal(t, tt.valid, valid)
			assert.Equal(t, tt.contentType, contentType)
		})
	}
}
